17.md

```diff
@@ -0,0 +1,111 @@
+
+**You need only know the nostr ID or bech32 form of a repo you will be able to discover all versions, its origins and its history.**
+
+The git object data is not designed to be stored in nostr, but rather, the signaling, update, version and commit data.  Git data will live in existing git repositories, github, gitlab, gitea, gogs, ssh, file system etc.
+
+
+New Kind
+========
+
+A special event with kind `17`, can be used to indicate git commit messages. This number was selected to be hopefully easy to remember : `G-1-7`.  
+
+
+New Tag
+=======
+
+A new tag "c" will be used to store commit hashes from git a new version of the repo 
+
+
+Bech32
+======
+
+The prefix `nrepo` can be used in front ends that support it, to allow easier copy and paste
+
+
+Event Structure
+===============
+
+```JSON
+{
+  "id": "32-bytes lowercase hex-encoded sha256 of the the serialized event data",
+  "pubkey": "32-bytes lowercase hex-encoded public key of the repo",
+  "created_at": "git_author_date",
+  "kind": 17,
+  "tags": [
+    ["c", "git hash", "recommended git URI"],
+    ["e", "previous commit id"]
+  ],
+  "content": "arbitrary string",
+  "sig": "64-bytes hex of the signature of the sha256 hash of the serialized event data, which is the same as the id field"
+}
+```
+
+Git events over nostr have the following attributes:
+
+**`created_at`** SHOULD be equal to git *author date*.  This is the same number format as nostr.
+
+**`tags`** MUST contain tag "c" which is equal to the hex-encoded git commit hash.  Typically this will be a 40 character sha-1 hash, but git will soon be allowing sha-2 hashes too
+
+MAY contain an entry identifying the previous commit, if one exists, in the form `["e", "<event_id>"]`.  Further tags can be developed to capture informative git information such as major/minor versioning.  TBD: it will be possible to hint at a recommended git URL to download the repos (e.g. from github), but it is not yet decided whether this would be a tag itself or an additional entry in an existing tag.
+
+**`pubkey`** SHOULD be associated with a given repo.  This could be communicated out of band, or, for example MAY be included in a file in the root directory of the repo, `nostr.json`
+
+**`content`** empty, reserved for future use
+
+Repo Discovery
+===============
+
+It will also be possible signal a recommended repo, just like it is possible to recommended a relay. In this way, a codebase could start on github, move to gitlab, then to an own hosted instance, and finally back to github.  All, without any breaks in service.  TODO: flesh out this section, perhaps include relay discovery, and bech32 encoding
+
+
+nostr.json
+==========
+
+The design of this NIP is such that each repo will have a public key used to send updates to a relay.  That key SHOULD appear in the top level directory of the repo in a file called `nostr.json`
```

Comment:
So if I fork a project, I should change that nostr.json to use my pubkey ... but not really my pubkey as I would have to create a pubkey for the project? And then I would replicate all the events of the upstream project using my key?

Reply:
This is a very good question.  Beyond the scope for now.  But it's entirely possibly, and I expect, that forked projects will have a new pubkey.  We will see!